build/result.go

```diff
@@ -90,18 +90,35 @@ func getResultAt(ctx context.Context, c *client.Client, solveOpt client.SolveOpt
 				}
 			}
 			// Record the client and ctx as well so that containers can be created from the SolveError.
+			doneCh := make(chan struct{})
 			resultCtx.res = res2
 			resultCtx.gwClient = c
 			resultCtx.gwCtx = ctx
-			resultCtx.gwDone = cancel
+			resultCtx.gwDone = func() {
+				close(doneCh)
+				// wait for Build() completion(or timeout) to ensure the Build's finalizing and avoiding an error "context canceled"
+				select {
+				case <-buildDoneCh:
+				case <-time.After(5 * time.Second):
```

Comment:
I don't think we should have a delay here, I think this should block?

Actually, instead of introducing another channel, could we keep the custom context we had before but use `context.WithCancelCause` and then use a custom error type to cancel it in the implementation of `gwDone`, then we can just detect that error when loading from `ctx.Err()` and return a `nil` error in that case? I think that might be a neater implementation that doesn't require we go and add another channel.

With the current refactor `gwCtx` isn't cancelled after `gwDone` is called, so we need to make sure it is.

Reply:
> > I'm not sure why we get different results between the local and remote controller? I'm guessing one gets cancelled, and the other doesn't?
> 
> Did you manage to work this out?

With this patch, the behaviour of both of the controllers seem to be the same. On `--sbom=false`, they successfully cleanup without "context canceled" error. On `--sbom=true`, they cause panic (https://github.com/docker/buildx/pull/1640#issuecomment-1513000048) on buildkitd.

> I don't think we should have a delay here, I think this should block?

BuildKit's `client.Build()` seems to block indefinitely when buildkitd panics. So timeout is needed if we want to avoid buildx blocking indefinitely when buildkitd panics.

> context.WithCancelCause

Fixed to use `context.WithCancelCause`. We still need a channel to ensure `gwDone` returns after `Build()` returns. Otherwise buildx process (running local controller) can exit after completion of `gwDone` but before completion of `Build()`, which results in BuildKit reporting cancellation error.

Reply:
Yup, this seems better to me, thanks :tada: